src/Lib.P2P/Routing/DistributedQuery.cs

```diff
@@ -0,0 +1,273 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Common.Logging;
+using MultiFormats;
+using ProtoBuf;
+
+namespace Lib.P2P.Routing
+{
+    /// <summary>
+    ///   A query that is sent to multiple peers.
+    /// </summary>
+    /// <typeparam name="T">
+    ///  The type of answer returned by a peer.
+    /// </typeparam>
+    public class DistributedQuery<T> where T : class
+    {
+        private static ILog log = LogManager.GetLogger("Lib.P2P.Routing.DistributedQuery");
+        private static int nextQueryId = 1;
+
+        /// <summary>
+        ///   The maximum number of peers that can be queried at one time
+        ///   for all distributed queries.
+        /// </summary>
+        private static SemaphoreSlim askCount = new SemaphoreSlim(128);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [A static field in a generic type is not shared among instances of different close constructed types.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)